torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -443,20 +436,39 @@ def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
                 from a call to :meth:`state_dict`
         """
 
+        # Update the state, trusting the ordering in param_groups
+        # Apart from the removal of states not owned by this rank, the pytorch logic is kept
+        # (See torch.optim.optimizer)
+        id_map = {
+            old_id: p
+            for old_id, p in zip(
+                chain.from_iterable((g["params"] for g in state_dict["param_groups"])),
+                chain.from_iterable((g["params"] for g in self.param_groups)),
+            )
+        }
+
+        _param_list = list(chain.from_iterable((g["params"] for g in self.param_groups)))
+
         for key, value in state_dict["state"].items():
-            param = self._index_to_param[key]
```

Comment:
this indexing was broken for interleaved param groups actually, my bad. The proposed changes are based on the default pytorch implementation (+minor changes to accommodate the sharding) and solved the PySpeech use case which was previously broken.